### DISCUSSION TITLE
infinite timer for Hello & TLU time interval

### DISCUSSION THREAD

**Discussion by xhaktm3** [ANLAB-KAIST/KENSv3#233] [Project Q&A 3. Internet Layer (IP/Routing)] [answered]
Hello TA, thank you for providing newly made PA#4 which is different from before semeters's RIPv1. Having fun doing PA, I got several questions about timer implementaton. 

In instruction, it has to multicast each type's message for each time intervals.
> For the Type field, you will implement only the following two types:
(1) - Hello: a message to discover neighboring routers. It is multicast to 224.0.0.5 every helloInt time interval.
(4) - Link State Update: a message containing all sender’s link costs. This message is sent upon a timer expiration in this assignment or forwarded to other routers. It is multicast to 224.0.0.5 every LSUInt time interval.

But, when i have to stop to send more messages and end the program? I implemented this multicast by calling addtimer() at the last part of timerCallBack, so it makes infinite timer call and test cannot be ended.

Without implementing the Hello&TLU time interval multicast, I got passed all testcases in testpwosf.cpp by mulitcasting every LSU call immediately and no hello multicast more than twice. In this case, can i get full score by just passing all test cases?

**Comment by Pusnow** [comment, marked as answer]
> But, when i have to stop to send more messages and end the program?

You can use `finalize()` method to cancel timers. However, it would not be your problem.
The reason of infinite test is probably routing loop.
A incorrect implementation of the routing protocol will cause packets to keep traveling through router cycles, increasing simulation time exponentially.
To avoid loops, you need to make right decisions about when to forward packets.


> Without implementing the Hello&TLU time interval multicast, I got passed all testcases in testpwosf.cpp by mulitcasting every LSU call immediately and no hello multicast more than twice. In this case, can i get full score by just passing all test cases?

Yes, you will get full score, but try to solve the infinite test run for your learning.
